### PR TITLE
simplifier from_elements and extract op pattern

### PR DESCRIPTION
**Pull request by Yancey1989**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc [hunk 1]
```diff
@@ -345,6 +345,31 @@ struct BroadCastInDimOfReshapeOpCanonicalizationPattern
   }
 };
 
+template <typename OpTy>
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc:348]: Why using template here?

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc [hunk 2]
```diff
@@ -345,6 +345,31 @@ struct BroadCastInDimOfReshapeOpCanonicalizationPattern
...
+                                PatternRewriter& rewriter) const override {
+    auto loc = op->getLoc();
+    Value input = op->getOperand(0);
+    Value result = op->getResult(0);
+    auto extractOp = input.getDefiningOp();
+    if (!isa<tensor::ExtractOp>(extractOp)) {
+      return failure();
+    }
+    auto rankTy = op.getType().template dyn_cast<RankedTensorType>();
+    auto extractInput = extractOp->getOperand(0);
```
Comment (wyzero) [inline tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc:362]: What if  the source tensor of`tensor.extract` having `rank > 0`. For example, `tensor.extract %0[%c0, ...]: tensor<?x?xf32>`?

If you not want to process the other case, please check and early return.

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc [hunk 3]
```diff
@@ -345,6 +345,31 @@ struct BroadCastInDimOfReshapeOpCanonicalizationPattern
...
+    auto outTy = RankedTensorType::get({1}, rankTy.getElementType());
+    auto newResult =
+        rewriter.create<mhlo::ReshapeOp>(loc, outTy, extractInput).getResult();
+    for (auto user : result.getUsers()) {
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc]: done

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc (hunk 1 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc]: removed the unused template.

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc (hunk 2 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc:366]: this pass only deal with input tensor of `tensor.extract` with `rank == 0`, I added some limitation and return early.

**Review by wyzero — DISMISSED** [review]
generally LGTM, with some minor comments.

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc [hunk 4]
```diff
@@ -345,6 +345,38 @@ struct BroadCastInDimOfReshapeOpCanonicalizationPattern
...
+struct SimplifierFromElementsPattern
+    : public OpRewritePattern<tensor::FromElementsOp> {
+  using OpRewritePattern<tensor::FromElementsOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::FromElementsOp op,
+                                PatternRewriter& rewriter) const override {
+    auto loc = op->getLoc();
+    Value input = op->getOperand(0);
+    Value result = op->getResult(0);
+    auto extractOp = input.getDefiningOp();
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc]: done

File: tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc [hunk 5]
```diff
@@ -345,6 +345,38 @@ struct BroadCastInDimOfReshapeOpCanonicalizationPattern
...
+    if (!isa<tensor::ExtractOp>(extractOp)) return failure();
+
+    auto extractInput = extractOp->getOperand(0);
+    auto extractRankTy =
+        extractInput.getType().template dyn_cast<RankedTensorType>();
+    if (extractRankTy.getRank() != 0)
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc]: good point, updated this pr.

**Review by tanyokwok — APPROVED** [review]
LGTM

**Pull request merged**